howdoi/plugins/base.py

```diff
@@ -0,0 +1,166 @@
+import os
+import re
+import sys
+
+import appdirs
+
+from cachelib import FileSystemCache, NullCache
+
+from pyquery import PyQuery as pq
+
+
+
+
+# Handle imports for Python 2 and 3
+if sys.version < '3':
+    import codecs
+    from urllib import quote as url_quote
+    from urllib import getproxies
+    from urlparse import urlparse, parse_qs
+
+    # Handling Unicode: http://stackoverflow.com/a/6633040/305414
+    def u(x):
+        return codecs.unicode_escape_decode(x)[0]
+else:
+    from urllib.request import getproxies
+    from urllib.parse import quote as url_quote, urlparse, parse_qs
+
+    def u(x):
+        return x
+
+# rudimentary standardized 3-level log output
+
+
+def _print_err(x):
+    print("[ERROR] " + x)
+
+
+_print_ok = print  # noqa: E305
+
+
+def _print_dbg(x):
+    print("[DEBUG] " + x)  # noqa: E302
+
+
+CACHE_EMPTY_VAL = "NULL"
+CACHE_DIR = appdirs.user_cache_dir('howdoi')
+CACHE_ENTRY_MAX = 128
+
+if os.getenv('HOWDOI_DISABLE_CACHE'):
+    cache = NullCache()  # works like an always empty cache
+else:
+    cache = FileSystemCache(CACHE_DIR, CACHE_ENTRY_MAX, default_timeout=0)
+
+ANSWER_HEADER = u('{2}  Answer from {0} {2}\n{1}')
+STAR_HEADER = u('\u2605')
+CACHE_EMPTY_VAL = "NULL"
+NO_ANSWER_MSG = '< no answer given >'
+
+if os.getenv('HOWDOI_DISABLE_SSL'):  # Set http instead of https
+    SCHEME = 'http://'
+    VERIFY_SSL_CERTIFICATE = False
+else:
+    SCHEME = 'https://'
+    VERIFY_SSL_CERTIFICATE = True
+
+BLOCK_INDICATORS = (
+    'form id="captcha-form"',
+    'This page appears when Google automatically detects requests coming from your computer '
+    'network which appear to be in violation of the <a href="//www.google.com/policies/terms/">Terms of Service'
+)
+
+URL = os.getenv('HOWDOI_URL') or 'stackoverflow.com'
+
+SEARCH_URLS = {
+    'bing': SCHEME + 'www.bing.com/search?q=site:{0}%20{1}&hl=en',
+    'google': SCHEME + 'www.google.com/search?q=site:{0}%20{1}&hl=en',
+    'duckduckgo': SCHEME + 'duckduckgo.com/?q=site:{0}%20{1}&t=hj&ia=web'
+}
+
+
+class BasePlugin():
```

Comment:
Architecture thoughts:

```python
import enum.Enum as Enum

class Source(Enum):
   CLI = 0
   DISCORD_MESSAGE = 1

class ServiceConnectionError(Exception):
  pass

class Plugin:
    name = "UnimplementedPlugin"
    def __init__():
      pass
    def resolve(self):
      pass
    def extract(self, source, data): # raises an exception if this fails
      pass

class PluginStack(list):
   def __init__(self, plugins):
     super(PluginStack, self).__init__(plugins)
   def handle_input(self, source, data):
     tries = [ ]
     for plugin in self:
       try:
         plugin.extract(source, data)
         plugin.resolve()
       except <any of the relevant exceptions>:
         tries.append("howdoi could not get a response from {}".format(plugin.name))
```

Further possible areas for improvement:
 * adding `async` support.
 * "real-time" error handling – i.e. if the first plugin fails, howdoi immediately sends a response
 * optional "supports" attribute for `Plugin`s which is a list containing all the sources they support.

Reply:
I'm trying to understand `PluginStack`. Is it the manager of the plugins installed in howdoi?

Reply:
Yes, that's the idea. Whenver it receives a request it tries each plugin in turn. If a plugin fails, it moves on to the next plugin, otherwise it returns the result of that plugin. The code above isn't 100% complete.

Reply:
Thank you. I think the user experience is having the user specify the plugin they want to use with a flag along with their query (at least for now, Hajer is working on a model that automatically detects which plugin is the most appropriate). How would things change then?

Reply:
I think that there has to be a list of installed plugins and a way of resolving/calling them somehow. A class to handle this could still be useful, but it would be a bit different to the `PluginStack`.